core/src/main/java/org/apache/calcite/rel/rules/ExpandDisjunctionForJoinInputsRule.java

```diff
@@ -0,0 +1,388 @@
+    call.transformTo(newFilter);
+  }
+
+  private static void matchJoin(ExpandDisjunctionForJoinInputsRule rule, RelOptRuleCall call) {
+    Join join = call.rel(0);
+    RelBuilder builder = call.builder();
+
+    ExpandDisjunctionForJoinInputsHelper helper =
+        new ExpandDisjunctionForJoinInputsHelper(
+            join.getLeft().getRowType().getFieldCount(),
+            join.getRight().getRowType().getFieldCount(),
+            join.getJoinType().canPushLeftFromWithin(),
+            join.getJoinType().canPushRightFromWithin(),
+            builder,
+            rule.config.bloat());
+    Map<String, RexNode> expandResult = helper.expand(join.getCondition());
+
+    RexNode newCondition =
+        call.builder().and(
+            join.getCondition(),
+            expandResult.getOrDefault(helper.LEFT_KEY, builder.literal(true)),
+            expandResult.getOrDefault(helper.RIGHT_KEY, builder.literal(true)));
+    if (newCondition.equals(join.getCondition())) {
+      return;
+    }
+    Join newJoin =
+        join.copy(
+            join.getTraitSet(),
+            newCondition,
+            join.getLeft(),
+            join.getRight(),
+            join.getJoinType(),
+            join.isSemiJoinDone());
+    call.transformTo(newJoin);
+  }
+
+  /**
+   * Helper class to expand predicates.
+   */
+  private static class ExpandDisjunctionForJoinInputsHelper {
+
+    private final ImmutableBitSet leftBitmap;
+
+    private final ImmutableBitSet rightBitmap;
+
+    private final boolean canPushLeft;
+
+    private final boolean canPushRight;
+
+    private final RelBuilder relBuilder;
+
+    private final int maxNodeCount;
+
+    // Used to record the number of redundant expressions expanded.
+    private int currentCount;
+
+    private static final String LEFT_KEY = "left";
+
+    private static final String RIGHT_KEY = "right";
+
+    private ExpandDisjunctionForJoinInputsHelper(
+        int leftFieldCount,
+        int rightFieldCount,
+        boolean canPushLeft,
+        boolean canPushRight,
+        RelBuilder relBuilder,
+        int maxNodeCount) {
+      this.leftBitmap = ImmutableBitSet.range(0, leftFieldCount);
+      this.rightBitmap = ImmutableBitSet.range(leftFieldCount, leftFieldCount + rightFieldCount);
+      this.canPushLeft = canPushLeft;
+      this.canPushRight = canPushRight;
+      this.relBuilder = relBuilder;
+      this.maxNodeCount = maxNodeCount;
+    }
+
+    private Map<String, RexNode> expand(RexNode condition) {
+      try {
+        this.currentCount = 0;
+        return expandDeep(condition);
+      } catch (OverLimitException e) {
+        return new HashMap<>();
+      }
+    }
+
+    /**
+     * Expand predicates recursively that can be pushed down to Join inputs.
+     *
+     * @param condition   Predicate to be expanded
+     * @return  A map from Join left/right input to a (combined) predicate that can be pushed down
+     * and only depends on columns of the join inputs.
```

Comment:
Maybe the extra "one" makes it more precise?

```suggestion
     * and only depends on columns of one the join inputs.
```